lib/dl_auth_api_lib/dl_auth_api_lib/settings.py

```diff
@@ -0,0 +1,56 @@
+from typing import (
+    Annotated,
+    Any,
+    Optional,
+)
+
+import pydantic
+import pydantic_settings
+
+
+class BaseOAuthClient(pydantic.BaseModel):
+    conn_type: str
+    auth_type: str
+
+    @classmethod
+    def factory(cls, data: Any) -> "BaseOAuthClient":
+        if isinstance(data, BaseOAuthClient):
+            return data
+
+        assert isinstance(data, dict), "OAuthClient settings must be a dict"
+        assert "auth_type" in data, "No auth_type in client"
+        assert data["auth_type"] in _REGISTRY, f"No such OAuth type: {data['auth_type']}"
+
+        config_class = _REGISTRY[data["auth_type"]]
+        return config_class.model_validate(data)
+
+    def as_dict(self) -> dict[str, str | None]:
+        raise NotImplementedError
+
+
+class YandexOAuthClient(BaseOAuthClient):
+    auth_type: str = "yandex"
+
+    client_id: str
+    client_secret: str
+    redirect_uri: str
+    scope: Optional[str] = pydantic.Field(default=None)
+
+    def as_dict(self) -> dict[str, str | None]:
+        return dict(
+            client_id=self.client_id, client_secret=self.client_secret, redirect_uri=self.redirect_uri, scope=self.scope
+        )
+
+
+class AuthAPISettings(pydantic_settings.BaseSettings):
+    model_config = pydantic_settings.SettingsConfigDict(env_nested_delimiter="__")
+
+    auth_clients: Annotated[
+        dict[str, pydantic.SerializeAsAny[BaseOAuthClient]],
+        pydantic.BeforeValidator(lambda v: {name: BaseOAuthClient.factory(data) for name, data in v.items()}),
```

Comment:
let's move out as dict_factory(factory) which makes dict factory out of normal one